### ISSUE TITLE
Read data from the packages?

### ISSUE THREAD

**Issue by phildini** [phildini/thanks#7]
A suggestion from @EricaJoy was to figure out some way of reading this info from the packages themselves, rather than a central database.

(Relates to https://github.com/feross/thanks/issues/2)

Today, few-to-no package authors are capturing this metadata in their packages. Two approaches that come to mind:

1. Make it possible for that info to go in the `setup.py`
    * Ideally this would also be fetchable via PyPI API.
2. Encourage maintainers to ship a .thanks.json in their packages, and check from the package to read from there.

Edit: I have filed https://github.com/pypa/setuptools/issues/1276 to try and get the ball rolling on this.

**Comment by kootenpv** [comment]
It somehow seems like a bad idea to have this centralized. 

How about someone evil sending a payment request to all people mentioned in this single `thanks.json` file?

**Comment by phildini** [comment]
So, let's be clear: Today, this needs to be centralized. It's going to be centralized until there's another way. Anyone who wants their info removed, I will respect - so far all of these have been volunteered or are on the public web.

This thread is for discussion about _how_ to get to decentralization.

**Comment by kootenpv** [comment]
I understand, I just came to raise an issue and it already fit in this thread.

I would vote for an extended CONTRIBUTORS file including payments info.

**Comment by untitaker** [comment]
If you are keeping a central database, you have to ensure that the PRs adding to that DB are not potentially malicious (i.e. somebody trying to funnel donations of a larger project to their own accounts). I don't think this moderation effort scales even in the mid term.

Both the scaling and the security problems go away if you tell packagers to add metadata to their own packages. They have plenty of incentive to do so.

**Comment by phildini** [comment]
@untitaker I agree, and I am in the process of filing issues with the relevant repos to make this easier  for all.

If you have suggestions of the right place to store it, I'm all ears.

**Comment by kootenpv** [comment]
I think an extended CONTRIBUTORS file including payment info would work

**Comment by phildini** [comment]
Update on where I'm thinking of heading. The spec for metadata behind `setup.py` already defines the capability for a `project_urls` dict to be an argument to setup.py. I'm thinking of encouraging maintainers to add a `Funding` url or something of the like. You can look at `https://github.com/phildini/thanks/commit/a4e549338eb3e3c70b1dd5628b38dcbdbf63443a` as an example.

**Comment by untitaker** [comment]
@kootenpv there is no standardized format for CONTRIBUTORS. I also wasn't under the impression that `thanks` was about supporting every contributor, but rather the (core) maintenance.

@phildini Apart from a field people have to add, what about scraping the README/long_description for patreon/flattr/paypal/...-links or simply links with "donation"/"donate" in them?

**Comment by phildini** [comment]
I'd rather not assume from data parsing if I can avoid it. Scraping a text file is not my ideal here, because if I guess wrong it's a crappy experience. 

As for supporting all contributors vs. supporting the project: Ideally I'd like to allow for both. The link in `setup.py` is a step towards getting the project supported, but both would be great.

**Comment by phildini** [comment]
Starting point:

```
[line for line in pkg_resources.get_distribution('thanks').get_metadata_lines(pkg_resources.get_distribution('thanks').PKG_INFO) if line.startswith('Project-URL:')]
```
